notebooks/k_nearest_neighbors.livemd

```diff
@@ -0,0 +1,833 @@
+```
+
+```elixir
+best_neighbors = Nx.add(Nx.argmin(res), min_num_neighbors) |> Nx.to_number()
+```
+
+The best value is 2. Now try this value to a final prediction.
+
+```elixir
+best_model =
+  KNearestNeighbors.fit(train_data, train_labels,
+    num_neighbors: best_neighbors,
+    num_classes: 3,
+    task: :regression
+  )
+
+final_prediction = KNearestNeighbors.predict(best_model, test_data)
+Scholar.Metrics.mean_absolute_error(test_labels, final_prediction) |> Nx.sqrt()
+```
+
+### Outliers prediction
+
+<!-- livebook:{"break_markdown":true} -->
+
+KNNs can be used for outlier detection. In this task, we will also use the Airfoil Self-Noise dataset. We will calculate for each sample a mean value of distances to the three closest points from dataset.
+
+```elixir
+model = KNearestNeighbors.fit(x, y, num_neighbors: 3, task: :regression)
+{num_samples, _} = Nx.shape(x)
+{distances, indices} = KNearestNeighbors.k_neighbors(model, x)
+
+mean_distances = Nx.mean(distances, axes: [1], keep_axes: true)
+
+samples_mean_distances =
+  Nx.concatenate([mean_distances, Nx.iota({num_samples, 1})], axis: 1)
+  |> DF.new()
+  |> DF.rename(["mean_distances", "index"])
+```
+
+Now, plot the mean values of distances.
+
+```elixir
+Vl.new(
+  title: [
+    text: "Plot to check anomalies",
+    offset: 20
+  ],
+  width: 800,
+  height: 630
+)
+|> Vl.layers([
+  Vl.new()
+  |> Vl.data_from_values(samples_mean_distances)
+  |> Vl.mark(:line, color: :jet)
+  |> Vl.encode_field(:x, "index", type: :quantitative, scale: [domain: [0, 1510]])
+  |> Vl.encode_field(:y, "mean_distances",
+    type: :quantitative,
+    title: "Mean distance to clusters",
+    scale: [domain: [0.0065, 0.012]]
+  ),
+  Vl.new()
+  |> Vl.data_from_values(DF.new(y: [0.009]))
+  |> Vl.mark(:rule, color: :red, stroke_dash: [5, 5])
+  |> Vl.encode_field(:y, "y", type: :quantitative, title: "")
+])
+```
+
+We can assume, that samples with average distance bigger than 0.009 can be treated as outliers.
+
+```elixir
+data = DF.put(data, :is_outlier, Nx.greater(mean_distances, 0.009))
+```
+
+```elixir
+Vl.new(
+  title: [
+    text: "Logarithm of Frequency vs Logarithm of Suction Side Displacement Thickness",
+    offset: 20
+  ],
+  width: 630,
+  height: 630
+)
+|> Vl.data_from_values(data)
+|> Vl.layers([
+  Vl.new()
+  |> Vl.param("brush", select: :interval)
+  |> Vl.mark(:circle)
+  |> Vl.encode_field(:x, "frequency_log", type: :quantitative, scale: [domain: [4.5, 10.5]])
+  |> Vl.encode_field(:y, "suction_side_displacement_thickness_log",
+    type: :quantitative,
+    scale: [domain: [-8.0, -2.5]]
+  )
+  |> Vl.encode(:color,
+    condition: [param: "brush", field: "is_outlier", type: :nominal],
+    value: :gray
+  ),
+  Vl.new()
+  |> Vl.transform(filter: [param: "brush", field: "is_outlier", equal: true])
+  |> Vl.mark(:text)
+  |> Vl.encode(:x, value: 11)
+  |> Vl.encode(:y, value: -3.5)
+  |> Vl.encode_field(:text, "filter", type: :quantitative, aggregate: :count)
+])
+```
```

Comment:
I'm wondering how to implement a plot in which there will be a counter of outliers inside a brush. Currently, the counter shows all samples inside a brush.